### PR TITLE
CStr Invariant proofs for bytes, as_ptr, to_str

### PR DESCRIPTION
**Pull request by rajathkotyal**
Towards #150 

Verifies that the CStr safety invariant holds after calling :

`bytes	core::ffi::c_str`
`to_str	core::ffi::c_str`
`as_ptr	core::ffi::c_str`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ffi/c_str.rs [hunk 1]
```diff
@@ -860,7 +860,7 @@ impl FusedIterator for Bytes<'_> {}
 mod verify {
     use super::*;
 
-    // pub const fn from_bytes_until_nul(bytes: &[u8]) -> Result<&CStr, FromBytesUntilNulError>
+    // pub const fn from_bytes_until_nul(bytes: &[u8]) -> Result<&CStr, FromBytesUntilNulError>
```
Comment (tautschnig) [inline library/core/src/ffi/c_str.rs:863]:
```suggestion
    // pub const fn from_bytes_until_nul(bytes: &[u8]) -> Result<&CStr, FromBytesUntilNulError>
```

**Review by tautschnig — APPROVED** [review]

File: library/core/src/ffi/c_str.rs [hunk 2]
```diff
@@ -875,4 +875,72 @@ mod verify {
...
+            let ptr = c_str.as_ptr();
+            let len = c_str.to_bytes_with_nul().len();
+
+            // We ensure that `ptr` is valid for reads of `len` bytes
+            unsafe {
+                for i in 0..len {
+                    // Iterate and get each byte in the C string from our raw ptr
+                    let byte_at_ptr = *ptr.add(i);
+                    // Get the byte at every pos from the to_bytes_with_nul method
+                    let byte_in_cstr = c_str.to_bytes_with_nul()[i];
```
[the remark this answers is not in the document]
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: fixed

**Comment by tautschnig** [comment]
@rajathkotyal Could you please resolve the conflicts that just came up?

**Comment by rajathmCMU** [comment]
Hi can we get a workflow approval for this? Thanks!

**Review by celinval — COMMENTED** [review]
Hi @rajathkotyal, is it possible to use `arbitrary_cstr` function in these harnesses?

**Comment by rajathkotyal** [comment]
Sure, updated code @celinval

**Review by celinval — APPROVED** [review]
Thanks!

**Comment by rajathkotyal** [comment]
Thanks @celinval , can we get a workflow approval for this PR?

**Pull request merged**